libsplinter/src/transport/ws/connection.rs

```diff
@@ -12,63 +12,55 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::io::{self, Read, Write};
-use std::os::unix::io::AsRawFd;
+use std::io::{Read, Write};
 
-use mio::{unix::EventedFd, Evented, Poll, PollOpt, Ready, Token};
-use websocket::{
-    client::sync::Client,
-    message::{Message, OwnedMessage::Binary},
-    result::WebSocketError,
-    stream::sync::{AsTcpStream, Stream},
-};
+use mio::Evented;
+use tungstenite::{protocol::WebSocket, Message};
 
 use crate::transport::{Connection, DisconnectError, RecvError, SendError};
 
-pub(super) struct WsClientConnection<S>
+pub(super) struct WsConnection<S>
 where
-    S: Read + Write + Send,
+    S: Read + Write + Send + Evented,
 {
-    client: Client<S>,
+    websocket: WebSocket<S>,
     remote_endpoint: String,
     local_endpoint: String,
 }
 
-impl<S> WsClientConnection<S>
+impl<S> WsConnection<S>
 where
-    S: Read + Write + Send,
+    S: Read + Write + Send + Evented,
 {
-    pub fn new(client: Client<S>, remote_endpoint: String, local_endpoint: String) -> Self {
-        WsClientConnection {
-            client,
+    pub fn new(websocket: WebSocket<S>, remote_endpoint: String, local_endpoint: String) -> Self {
+        WsConnection {
+            websocket,
             remote_endpoint,
             local_endpoint,
         }
     }
 }
 
-impl<S> Connection for WsClientConnection<S>
+impl<S> Connection for WsConnection<S>
 where
-    S: AsTcpStream + Stream + Read + Write + Send,
+    S: Read + Write + Send + Evented,
 {
     fn send(&mut self, message: &[u8]) -> Result<(), SendError> {
-        Ok(self.client.send_message(&Message::binary(message))?)
+        self.websocket
+            .write_message(Message::Binary(message.to_vec()))?;
+        self.websocket.write_pending()?;
+        Ok(())
     }
 
     fn recv(&mut self) -> Result<Vec<u8>, RecvError> {
-        match self.client.recv_message() {
+        match self.websocket.read_message() {
             Ok(message) => match message {
-                Binary(v) => Ok(v),
+                Message::Binary(v) => Ok(v),
                 _ => Err(RecvError::ProtocolError(
-                    "message received was not
-                        websocket::message::OwnedMessage::Binary"
-                        .to_string(),
+                    "message received was not binary".to_string(),
                 )),
             },
-            Err(WebSocketError::IoError(ref e)) if e.kind() == io::ErrorKind::WouldBlock => {
-                Err(RecvError::WouldBlock)
-            }
-            Err(WebSocketError::NoDataAvailable) => Err(RecvError::WouldBlock),
+            Err(tungstenite::error::Error::Io(e)) => Err(RecvError::from(e)),
```

Comment:
I think your `From<tungstenite::error::Error>` covers all of these errors. You may be able to convert the error part of this match statement to a `?`.

Reply:
This piece was reworked a bit to cut down on the duplication.